tower-buffer/src/lib.rs

```diff
@@ -202,9 +199,7 @@ where
                         state: ResponseState::Failed(self.get_error_on_closed()),
                     }
                 } else {
-                    ResponseFuture {
-                        state: ResponseState::Full,
-                    }
+                    panic!("buffer full; poll_ready must be called first");
```

Comment:
I know in Linkerd2, we make use of buffers in places where calling `poll_ready` isn't feasible or possible. This change would crash the proxy instead of denying the request.

Reply:
Ah yes, linkerd used this strategy before we came up w/ a good alternative. The strategy for linkerd is to clone the request into the response future (I believe the router is the primary spot for this).

Reply:
Clone the request? Or clone the `Service`?

Reply:
Another possible pitfall: there isn't a guarantee that `poll_ready` means that `call` could succeed. If the service is backed by a channel, and a user calls `poll_ready` and in-between being notified it is ready, the receiver could be dropped. There'd be no way to guard against the panic in that case...

Reply:
poll_ready returning `Ready` does not mean that the call will succeed. It is fine for `poll_ready` to return `Ready` and `call` to fail the request.

Reply:
I was thinking of the behavior of `futures::sync`, where just because a sender saw ready from `poll_ready` doesn't mean a different sender couldn't use the slot and fill the buffer first. tokio-sync's channel will reserve the slot, so this can't happen.

What do you think of adding a comment here explaining this, so we know why this panic is justified?

Reply:
I added a comment.